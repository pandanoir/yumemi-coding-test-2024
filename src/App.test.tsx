// @vitest-environment jsdom

import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  test,
  vi,
} from 'vitest';
import { act, render, screen } from '@testing-library/react';
import { setupServer } from 'msw/node';
import { http, HttpResponse } from 'msw';
import type { JSX } from 'react';
import ResizeObserver from 'resize-observer-polyfill';

import prefecturesFixture from '../test/fixtures/prefectures.json';
import population47Fixture from '../test/fixtures/population47.json';

const handlers = [
  http.get(
    'https://yumemi-frontend-engineer-codecheck-api.vercel.app/api/v1/prefectures',
    () => HttpResponse.json(prefecturesFixture),
  ),
  http.get(
    'https://yumemi-frontend-engineer-codecheck-api.vercel.app/api/v1/population/composition/perYear?prefCode=47',
    () => HttpResponse.json(population47Fixture),
  ),
];

const server = setupServer(...handlers);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

window.ResizeObserver ??= ResizeObserver;

describe('App', () => {
  let App: () => JSX.Element;
  beforeAll(async () => {
    App = (await import('./App')).App;
  });
  beforeEach(() => {
    vi.resetModules(); // 各テストの前にモジュールキャッシュをリセット
  });

  test('snapshot', async () => {
    const { asFragment } = await act(() => render(<App />));
    expect(asFragment()).toMatchSnapshot();
  });

  it('は選択された都道府県の折れ線グラフを表示する', async () => {
    const { asFragment } = await act(() => render(<App />));
    await act(async () => {
      screen.getByRole('checkbox', { name: '沖縄県' }).click();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('は選択解除された都道府県の折れ線グラフを表示しない', async () => {
    const { asFragment } = await act(() => render(<App />));
    await act(async () => {
      screen.getByRole('checkbox', { name: '沖縄県' }).click();
    });
    await act(async () => {
      screen.getByRole('checkbox', { name: '沖縄県' }).click();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('はAPIサーバーが落ちていた場合になにも表示しない', async () => {
    // Appは読み込んだ時点でfetchを開始するので、サーバーレスポンスをエラーに設定してからdynamic importしている
    server.use(
      http.get(
        'https://yumemi-frontend-engineer-codecheck-api.vercel.app/api/v1/prefectures',
        () => HttpResponse.error(),
      ),
    );
    const App = (await import('./App')).App;
    const { asFragment } = await act(() => render(<App />));
    expect(asFragment()).toMatchSnapshot();
  });

  it('は選択した人口を表示できる', async () => {
    const { asFragment } = await act(() => render(<App />));
    await act(async () => {
      screen.getByRole('option', { name: '生産年齢人口' }).click();
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
