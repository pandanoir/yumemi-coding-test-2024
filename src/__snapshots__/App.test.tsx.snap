// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > snapshot 1`] = `
<DocumentFragment>
  <div>
    <select>
      <option>
        総人口
      </option>
      <option>
        年少人口
      </option>
      <option>
        生産年齢人口
      </option>
      <option>
        老年人口
      </option>
    </select>
    <br />
    <div
      class="prefecture-checkbox-list"
    >
      <label>
        <input
          type="checkbox"
        />
        北海道
      </label>
      <label>
        <input
          type="checkbox"
        />
        青森県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岩手県
      </label>
      <label>
        <input
          type="checkbox"
        />
        宮城県
      </label>
      <label>
        <input
          type="checkbox"
        />
        秋田県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山形県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        茨城県
      </label>
      <label>
        <input
          type="checkbox"
        />
        栃木県
      </label>
      <label>
        <input
          type="checkbox"
        />
        群馬県
      </label>
      <label>
        <input
          type="checkbox"
        />
        埼玉県
      </label>
      <label>
        <input
          type="checkbox"
        />
        千葉県
      </label>
      <label>
        <input
          type="checkbox"
        />
        東京都
      </label>
      <label>
        <input
          type="checkbox"
        />
        神奈川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        新潟県
      </label>
      <label>
        <input
          type="checkbox"
        />
        富山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        石川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福井県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山梨県
      </label>
      <label>
        <input
          type="checkbox"
        />
        長野県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岐阜県
      </label>
      <label>
        <input
          type="checkbox"
        />
        静岡県
      </label>
      <label>
        <input
          type="checkbox"
        />
        愛知県
      </label>
      <label>
        <input
          type="checkbox"
        />
        三重県
      </label>
      <label>
        <input
          type="checkbox"
        />
        滋賀県
      </label>
      <label>
        <input
          type="checkbox"
        />
        京都府
      </label>
      <label>
        <input
          type="checkbox"
        />
        大阪府
      </label>
      <label>
        <input
          type="checkbox"
        />
        兵庫県
      </label>
      <label>
        <input
          type="checkbox"
        />
        奈良県
      </label>
      <label>
        <input
          type="checkbox"
        />
        和歌山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        鳥取県
      </label>
      <label>
        <input
          type="checkbox"
        />
        島根県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岡山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        広島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山口県
      </label>
      <label>
        <input
          type="checkbox"
        />
        徳島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        香川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        愛媛県
      </label>
      <label>
        <input
          type="checkbox"
        />
        高知県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福岡県
      </label>
      <label>
        <input
          type="checkbox"
        />
        佐賀県
      </label>
      <label>
        <input
          type="checkbox"
        />
        長崎県
      </label>
      <label>
        <input
          type="checkbox"
        />
        熊本県
      </label>
      <label>
        <input
          type="checkbox"
        />
        大分県
      </label>
      <label>
        <input
          type="checkbox"
        />
        宮崎県
      </label>
      <label>
        <input
          type="checkbox"
        />
        鹿児島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        沖縄県
      </label>
    </div>
    <div
      style="opacity: 1;"
    >
      表示したい都道府県を選択してください
    </div>
  </div>
</DocumentFragment>
`;

exports[`App > はAPIサーバーが落ちていた場合になにも表示しない 1`] = `
<DocumentFragment>
  error!
</DocumentFragment>
`;

exports[`App > は選択された都道府県の折れ線グラフを表示する 1`] = `
<DocumentFragment>
  <div>
    <select>
      <option>
        総人口
      </option>
      <option>
        年少人口
      </option>
      <option>
        生産年齢人口
      </option>
      <option>
        老年人口
      </option>
    </select>
    <br />
    <div
      class="prefecture-checkbox-list"
    >
      <label>
        <input
          type="checkbox"
        />
        北海道
      </label>
      <label>
        <input
          type="checkbox"
        />
        青森県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岩手県
      </label>
      <label>
        <input
          type="checkbox"
        />
        宮城県
      </label>
      <label>
        <input
          type="checkbox"
        />
        秋田県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山形県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        茨城県
      </label>
      <label>
        <input
          type="checkbox"
        />
        栃木県
      </label>
      <label>
        <input
          type="checkbox"
        />
        群馬県
      </label>
      <label>
        <input
          type="checkbox"
        />
        埼玉県
      </label>
      <label>
        <input
          type="checkbox"
        />
        千葉県
      </label>
      <label>
        <input
          type="checkbox"
        />
        東京都
      </label>
      <label>
        <input
          type="checkbox"
        />
        神奈川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        新潟県
      </label>
      <label>
        <input
          type="checkbox"
        />
        富山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        石川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福井県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山梨県
      </label>
      <label>
        <input
          type="checkbox"
        />
        長野県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岐阜県
      </label>
      <label>
        <input
          type="checkbox"
        />
        静岡県
      </label>
      <label>
        <input
          type="checkbox"
        />
        愛知県
      </label>
      <label>
        <input
          type="checkbox"
        />
        三重県
      </label>
      <label>
        <input
          type="checkbox"
        />
        滋賀県
      </label>
      <label>
        <input
          type="checkbox"
        />
        京都府
      </label>
      <label>
        <input
          type="checkbox"
        />
        大阪府
      </label>
      <label>
        <input
          type="checkbox"
        />
        兵庫県
      </label>
      <label>
        <input
          type="checkbox"
        />
        奈良県
      </label>
      <label>
        <input
          type="checkbox"
        />
        和歌山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        鳥取県
      </label>
      <label>
        <input
          type="checkbox"
        />
        島根県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岡山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        広島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山口県
      </label>
      <label>
        <input
          type="checkbox"
        />
        徳島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        香川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        愛媛県
      </label>
      <label>
        <input
          type="checkbox"
        />
        高知県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福岡県
      </label>
      <label>
        <input
          type="checkbox"
        />
        佐賀県
      </label>
      <label>
        <input
          type="checkbox"
        />
        長崎県
      </label>
      <label>
        <input
          type="checkbox"
        />
        熊本県
      </label>
      <label>
        <input
          type="checkbox"
        />
        大分県
      </label>
      <label>
        <input
          type="checkbox"
        />
        宮崎県
      </label>
      <label>
        <input
          type="checkbox"
        />
        鹿児島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        沖縄県
      </label>
    </div>
    <div
      style="opacity: 1;"
    >
      <div
        class="graph-container"
      >
        <div>
          <div
            class="recharts-responsive-container"
            style="width: 100%; height: 100%; min-width: 0;"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App > は選択した人口を表示できる 1`] = `
<DocumentFragment>
  <div>
    <select>
      <option>
        総人口
      </option>
      <option>
        年少人口
      </option>
      <option>
        生産年齢人口
      </option>
      <option>
        老年人口
      </option>
    </select>
    <br />
    <div
      class="prefecture-checkbox-list"
    >
      <label>
        <input
          type="checkbox"
        />
        北海道
      </label>
      <label>
        <input
          type="checkbox"
        />
        青森県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岩手県
      </label>
      <label>
        <input
          type="checkbox"
        />
        宮城県
      </label>
      <label>
        <input
          type="checkbox"
        />
        秋田県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山形県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        茨城県
      </label>
      <label>
        <input
          type="checkbox"
        />
        栃木県
      </label>
      <label>
        <input
          type="checkbox"
        />
        群馬県
      </label>
      <label>
        <input
          type="checkbox"
        />
        埼玉県
      </label>
      <label>
        <input
          type="checkbox"
        />
        千葉県
      </label>
      <label>
        <input
          type="checkbox"
        />
        東京都
      </label>
      <label>
        <input
          type="checkbox"
        />
        神奈川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        新潟県
      </label>
      <label>
        <input
          type="checkbox"
        />
        富山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        石川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福井県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山梨県
      </label>
      <label>
        <input
          type="checkbox"
        />
        長野県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岐阜県
      </label>
      <label>
        <input
          type="checkbox"
        />
        静岡県
      </label>
      <label>
        <input
          type="checkbox"
        />
        愛知県
      </label>
      <label>
        <input
          type="checkbox"
        />
        三重県
      </label>
      <label>
        <input
          type="checkbox"
        />
        滋賀県
      </label>
      <label>
        <input
          type="checkbox"
        />
        京都府
      </label>
      <label>
        <input
          type="checkbox"
        />
        大阪府
      </label>
      <label>
        <input
          type="checkbox"
        />
        兵庫県
      </label>
      <label>
        <input
          type="checkbox"
        />
        奈良県
      </label>
      <label>
        <input
          type="checkbox"
        />
        和歌山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        鳥取県
      </label>
      <label>
        <input
          type="checkbox"
        />
        島根県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岡山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        広島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山口県
      </label>
      <label>
        <input
          type="checkbox"
        />
        徳島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        香川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        愛媛県
      </label>
      <label>
        <input
          type="checkbox"
        />
        高知県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福岡県
      </label>
      <label>
        <input
          type="checkbox"
        />
        佐賀県
      </label>
      <label>
        <input
          type="checkbox"
        />
        長崎県
      </label>
      <label>
        <input
          type="checkbox"
        />
        熊本県
      </label>
      <label>
        <input
          type="checkbox"
        />
        大分県
      </label>
      <label>
        <input
          type="checkbox"
        />
        宮崎県
      </label>
      <label>
        <input
          type="checkbox"
        />
        鹿児島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        沖縄県
      </label>
    </div>
    <div
      style="opacity: 1;"
    >
      表示したい都道府県を選択してください
    </div>
  </div>
</DocumentFragment>
`;

exports[`App > は選択解除された都道府県の折れ線グラフを表示しない 1`] = `
<DocumentFragment>
  <div>
    <select>
      <option>
        総人口
      </option>
      <option>
        年少人口
      </option>
      <option>
        生産年齢人口
      </option>
      <option>
        老年人口
      </option>
    </select>
    <br />
    <div
      class="prefecture-checkbox-list"
    >
      <label>
        <input
          type="checkbox"
        />
        北海道
      </label>
      <label>
        <input
          type="checkbox"
        />
        青森県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岩手県
      </label>
      <label>
        <input
          type="checkbox"
        />
        宮城県
      </label>
      <label>
        <input
          type="checkbox"
        />
        秋田県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山形県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        茨城県
      </label>
      <label>
        <input
          type="checkbox"
        />
        栃木県
      </label>
      <label>
        <input
          type="checkbox"
        />
        群馬県
      </label>
      <label>
        <input
          type="checkbox"
        />
        埼玉県
      </label>
      <label>
        <input
          type="checkbox"
        />
        千葉県
      </label>
      <label>
        <input
          type="checkbox"
        />
        東京都
      </label>
      <label>
        <input
          type="checkbox"
        />
        神奈川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        新潟県
      </label>
      <label>
        <input
          type="checkbox"
        />
        富山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        石川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福井県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山梨県
      </label>
      <label>
        <input
          type="checkbox"
        />
        長野県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岐阜県
      </label>
      <label>
        <input
          type="checkbox"
        />
        静岡県
      </label>
      <label>
        <input
          type="checkbox"
        />
        愛知県
      </label>
      <label>
        <input
          type="checkbox"
        />
        三重県
      </label>
      <label>
        <input
          type="checkbox"
        />
        滋賀県
      </label>
      <label>
        <input
          type="checkbox"
        />
        京都府
      </label>
      <label>
        <input
          type="checkbox"
        />
        大阪府
      </label>
      <label>
        <input
          type="checkbox"
        />
        兵庫県
      </label>
      <label>
        <input
          type="checkbox"
        />
        奈良県
      </label>
      <label>
        <input
          type="checkbox"
        />
        和歌山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        鳥取県
      </label>
      <label>
        <input
          type="checkbox"
        />
        島根県
      </label>
      <label>
        <input
          type="checkbox"
        />
        岡山県
      </label>
      <label>
        <input
          type="checkbox"
        />
        広島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        山口県
      </label>
      <label>
        <input
          type="checkbox"
        />
        徳島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        香川県
      </label>
      <label>
        <input
          type="checkbox"
        />
        愛媛県
      </label>
      <label>
        <input
          type="checkbox"
        />
        高知県
      </label>
      <label>
        <input
          type="checkbox"
        />
        福岡県
      </label>
      <label>
        <input
          type="checkbox"
        />
        佐賀県
      </label>
      <label>
        <input
          type="checkbox"
        />
        長崎県
      </label>
      <label>
        <input
          type="checkbox"
        />
        熊本県
      </label>
      <label>
        <input
          type="checkbox"
        />
        大分県
      </label>
      <label>
        <input
          type="checkbox"
        />
        宮崎県
      </label>
      <label>
        <input
          type="checkbox"
        />
        鹿児島県
      </label>
      <label>
        <input
          type="checkbox"
        />
        沖縄県
      </label>
    </div>
    <div
      style="opacity: 1;"
    >
      表示したい都道府県を選択してください
    </div>
  </div>
</DocumentFragment>
`;
